asynction/security.py

```diff
@@ -0,0 +1,391 @@
+
+    def http_api_key_security_check(request: Request):
+        api_key = None
+
+        if api_key_in == "query":
+            if api_key_name is not None:
+                api_key = request.args.get(api_key_name)
+        elif api_key_in == "header":
+            if api_key_name is not None:
+                api_key = request.headers.get(api_key_name)
+        elif api_key_in == "cookie":
+            cookies_list = request.headers.get("Cookie")
+            api_key = get_cookie_value(cookies_list, api_key_name)
+        else:
+            return None
+
+        if api_key is None:
+            return None
+
+        token_info = api_key_info_func(api_key, required_scopes, None)
+        if token_info is None:
+            return None
+
+        token_scopes = token_info.get("scope", token_info.get("scopes", ""))
+
+        if not scope_validate_func(required_scopes, token_scopes):
+            raise SecurityException(
+                f"Invalid scopes: required: {required_scopes}, provided: {token_scopes}"  # noqa: 501
+            )
+
+        return token_info
+
+    return http_api_key_security_check
+
+
+def build_oauth2_security_check(
+    requirement: SecurityRequirement, scheme: SecurityScheme
+) -> Callable[[Request], Mapping]:
+    token_info_func = load_token_info_func(scheme)
+    required_scopes = requirement.scopes
+
+    def oauth2_security_check(request: Request):
+        token_info = validate_authorization_header(request, token_info_func)
+        scope_validate_func = load_scope_validate_func(scheme)
+
+        if token_info is None:
+            return None
+
+        token_scopes = token_info.get("scope", token_info.get("scopes", ""))
+
+        if not scope_validate_func(required_scopes, token_scopes):
+            raise SecurityException(
+                f"Invalid scopes: required: {required_scopes}, provided: {token_scopes}"  # noqa: 501
+            )
+
+        return token_info
+
+    return oauth2_security_check
+
+
+# Dispatch table mapping SecuritySchemesType to security check builder
+_BUILDER_DISPATCH = {
+    SecuritySchemesType.HTTP: build_http_security_check,
+    SecuritySchemesType.OAUTH2: build_oauth2_security_check,
+    SecuritySchemesType.HTTP_API_KEY: build_http_api_key_security_check,
+}
+
+
+def build_security_handler(
+    security: Sequence[SecurityRequirement],
+    security_schemes: Mapping[str, SecurityScheme],
+) -> Callable[[Request], Mapping]:
+    # build a list of security validators based on the provided security schemes
+    security_checks = []
+
+    for requirement in security:
+        scheme = security_schemes.get(requirement.name)
+        if not scheme:
+            raise UnregisteredSecurityScheme(requirement.name)
+
+        scheme_type = SecuritySchemesType(scheme.type)
+        builder = _BUILDER_DISPATCH.get(scheme_type)
+        if not builder:
+            raise UnsupportedSecurityScheme(str(scheme_type))
```

Comment:
```suggestion
            continue
```
Asynction shouldn't be that strict. We should allow users to define/use not supported schemes, and they can just implement the relevant security within their connection handler that they provide as part of the spec YAML.